pgjdbc/src/main/java/org/postgresql/core/QueryExecutorCloseAction.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The action performs connection cleanup, so it is properly terminated from the backend
+ * point of view.
+ * Implementation note: it should keep only the minimum number of object references
+ * to reduce heap usage in case the user abandons connection without closing it first.
+ */
+public class QueryExecutorCloseAction implements Closeable {
```

Comment:
does this need to be public?
could this be final?

Reply:
Previously, `QueryExecutorBase.sendCloseMessage` was open, so I need to keep this class open